### PR TITLE
DMP-2029 - Add retention functionality to the StopAndCloseHandler

### PR DESCRIPTION
**Pull request by cmagowan-hmcts**
https://tools.hmcts.net/jira/browse/DMP-2029

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/darts/retention/helper/RetentionDateHelper.java [hunk 1]
```diff
@@ -45,25 +46,38 @@ public RetentionPolicyTypeEntity getRetentionPolicy(RetentionPolicyEnum policy)
...
 
-        int years = getValueFromPolicyArray(policyArray, "Y");
-        int months = getValueFromPolicyArray(policyArray, "M");
-        int days = getValueFromPolicyArray(policyArray, "D");
+        String[] policyArray = StringUtils.splitByCharacterType(StringUtils.trimToEmpty(policyString));
 
-        LocalDate newDate = dateToAppend.plusYears(years);
-        newDate = newDate.plusMonths(months);
-        newDate = newDate.plusDays(days);
+        LocalDate newDate = dateToAppend.plusYears(NumberUtils.toInt(policyArray[0]));
```
Comment (mestebanez) [inline src/main/java/uk/gov/hmcts/darts/retention/helper/RetentionDateHelper.java:59]: Avoid magic numbers in favour of more descriptive constants

File: src/main/java/uk/gov/hmcts/darts/retention/helper/RetentionDateHelper.java [hunk 2]
```diff
@@ -45,25 +46,38 @@ public RetentionPolicyTypeEntity getRetentionPolicy(RetentionPolicyEnum policy)
...
+            LocalDate policyDate = applyPolicyString(dateToAppend, retentionPolicyType.getDuration());
+            LocalDate termInRequest = applyPolicyString(dateToAppend, policyString);
+            return latestDate(policyDate, termInRequest);
+        }
+
+        return applyPolicyString(dateToAppend, retentionPolicyType.getDuration());
+    }
+
+    private LocalDate latestDate(LocalDate date1, LocalDate date2) {
+        if (date1.isAfter(date2)) {
```
Comment (mestebanez) [inline src/main/java/uk/gov/hmcts/darts/retention/helper/RetentionDateHelper.java:77]: ternary operation is better

File: src/main/java/uk/gov/hmcts/darts/event/service/handler/TranscriptionRequestHandler.java [hunk 3]
```diff
@@ -48,7 +50,7 @@ public void handle(final DartsEvent dartsEvent, EventHandlerEntity eventHandler)
         transcriptionRequestDetails.setTranscriptionTypeId(TranscriptionTypeEnum.OTHER.getId());
         transcriptionRequestDetails.setTranscriptionUrgencyId(TranscriptionUrgencyEnum.OVERNIGHT.getId());
         RequestTranscriptionResponse transcriptionResponse = transcriptionsApi.saveTranscriptionRequest(
-            transcriptionRequestDetails, false);
+                transcriptionRequestDetails, false);
```
Comment (mestebanez) [inline src/main/java/uk/gov/hmcts/darts/event/service/handler/TranscriptionRequestHandler.java:53]: sorry indention weirdness here

File: src/main/java/uk/gov/hmcts/darts/event/service/handler/SentencingRemarksAndRetentionPolicyHandler.java [hunk 4]
```diff
@@ -31,17 +32,18 @@ public SentencingRemarksAndRetentionPolicyHandler(RetrieveCoreObjectService retr
...
+                                                      AuthorisationApi authorisationApi) {
+        super(retrieveCoreObjectService, eventRepository, hearingRepository, caseRepository, eventPublisher, authorisationApi);
         this.transcriptionsApi = transcriptionsApi;
     }
 
     @Override
     @Transactional
     public void handle(final DartsEvent dartsEvent, EventHandlerEntity eventHandler) {
         var transcriptionRequestDetails = transcriptionRequestDetailsFrom(
-            dartsEvent,
```
Comment (mestebanez) [inline src/main/java/uk/gov/hmcts/darts/event/service/handler/SentencingRemarksAndRetentionPolicyHandler.java:43]: Do we want to add info/debug level logging across

File: src/main/java/uk/gov/hmcts/darts/common/repository/CaseRetentionRepository.java [hunk 5]
```diff
@@ -18,35 +19,64 @@ public interface CaseRetentionRepository extends JpaRepository<CaseRetentionEnti
...
     @Query("""
-        SELECT c
-        FROM CaseRetentionEntity c
-        WHERE c.courtCase = :courtCase
-        and c.retentionPolicyType.fixedPolicyKey not in ('PERM', 'MANUAL')
-        AND c.currentState='COMPLETE'
-        ORDER BY c.createdDateTime desc
-        limit 1
-        """
+            SELECT c
```
Comment (mestebanez) [inline src/main/java/uk/gov/hmcts/darts/common/repository/CaseRetentionRepository.java:22]: has anything changed here

File: src/integrationTest/java/uk/gov/hmcts/darts/event/service/impl/StandardEventHandlerTest.java [hunk 6]
```diff
@@ -34,39 +39,49 @@ class StandardEventHandlerTest extends IntegrationBaseWithGatewayStub {
...
+
     @Test
     void throwsOnUnknownCourthouse() {
         dartsDatabase.save(someMinimalCase());
         DartsEvent dartsEvent = someMinimalDartsEvent().courthouse(UNKNOWN_COURTHOUSE);
         dartsEvent.setCaseNumbers(List.of("123"));
         dartsEvent.setDateTime(today);
         assertThatThrownBy(() -> eventDispatcher.receive(dartsEvent))
-            .isInstanceOf(DartsApiException.class);
+                .isInstanceOf(DartsApiException.class);
```
Comment (mestebanez) [inline src/integrationTest/java/uk/gov/hmcts/darts/event/service/impl/StandardEventHandlerTest.java:59]: loads of formatting changes here? Can we clean these up hard to see the wood for the trees

File: src/main/java/uk/gov/hmcts/darts/retention/helper/RetentionDateHelper.java (hunk 1 above)
Comment (cmagowan-hmcts) [reply src/main/java/uk/gov/hmcts/darts/retention/helper/RetentionDateHelper.java]: done

File: src/main/java/uk/gov/hmcts/darts/retention/helper/RetentionDateHelper.java (hunk 2 above)
Comment (cmagowan-hmcts) [reply src/main/java/uk/gov/hmcts/darts/retention/helper/RetentionDateHelper.java]: done

File: src/main/java/uk/gov/hmcts/darts/event/service/handler/TranscriptionRequestHandler.java (hunk 3 above)
Comment (cmagowan-hmcts) [reply src/main/java/uk/gov/hmcts/darts/event/service/handler/TranscriptionRequestHandler.java]: i installed the new version of intellij, which seemed to change its standard indenting, i have now overrided that to specify the continuation indent to 4 in the .idea/codeStyles/Project.xml

File: src/integrationTest/java/uk/gov/hmcts/darts/event/service/impl/StandardEventHandlerTest.java (hunk 6 above)
Comment (cmagowan-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/darts/event/service/impl/StandardEventHandlerTest.java]: formatting fixed

File: src/main/java/uk/gov/hmcts/darts/common/repository/CaseRetentionRepository.java (hunk 5 above)
Comment (cmagowan-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/repository/CaseRetentionRepository.java]: formatting fixed

File: src/main/java/uk/gov/hmcts/darts/event/service/handler/SentencingRemarksAndRetentionPolicyHandler.java (hunk 4 above)
Comment (cmagowan-hmcts) [reply src/main/java/uk/gov/hmcts/darts/event/service/handler/SentencingRemarksAndRetentionPolicyHandler.java]: i wasn't actually changing this handler, i just changed the constructor, that affected all handlers.

**Review by mestebanez — APPROVED** [review]

**Pull request merged**